ibis/backends/base/__init__.py

```diff
@@ -136,17 +136,62 @@ class BaseBackend(abc.ABC):
     database_class = Database
     table_class: type[ops.DatabaseTable] = ops.DatabaseTable
 
+    def __init__(self, *args, **kwargs):
+        self._con_args: tuple[Any] = args
+        self._con_kwargs: dict[str, Any] = kwargs
+
+    def __getstate__(self):
+        return dict(
+            database_class=self.database_class,
+            table_class=self.table_class,
+            _con_args=self._con_args,
+            _con_kwargs=self._con_kwargs,
+        )
+
+    def __hash__(self):
+        return hash(self.datasource)
+
+    def __eq__(self, other):
+        return self.datasource == other.datasource
+
     @property
     @abc.abstractmethod
     def name(self) -> str:
         """
         Name of the backend, for example 'sqlite'.
         """
 
+    @property
+    def datasource(self) -> str:
+        """
+        Identity of the database.  Multiple connections to the same
+        database will have the same datasource.  Default implementation
+        assumes connection parameters uniquely specify the database.
+        """
+        parts = [self.table_class.__name__]
+        parts.extend(self._con_args)
+        parts.extend(f'{k}={v}' for k, v in self._con_kwargs.items())
+        return '_'.join(map(str, parts))
+
+    def connect(self, *args, **kwargs):
+        """
+        Return new client object with saved args/kwargs, having called
+        .reconnect() on it.
+        """
+        new_backend = self.__class__(*args, **kwargs)
+        new_backend.reconnect()
+        return new_backend
+
+    def reconnect(self):
+        """
+        Reconnect to the target database already configured with connect().
+        """
+        self.do_connect(*self._con_args, **self._con_kwargs)
+
     @abc.abstractmethod
-    def connect(connection_string, **options):
```

Comment:
can we preserve this for backwards compatibilty

Reply:
If you look a few lines above this, it's still there, so you can continue to use `connect()` like you always have.

Reply:
ahh fi that's the case, then just document connect & db_connect a bit

Reply:
Okay, I'm not clear on what's missing, or where this documentation would go.  Backends need to override the abstractmethod `do_connect`, but that's virtually identical to what they had to do previously with `connect` (as shown in [this comment in the PR](https://github.com/ibis-project/ibis/pull/3127#issuecomment-977467918)).  What else do you need?